runtime/near-vm-logic/src/logic.rs

```diff
@@ -2786,6 +2786,37 @@ impl<'a> VMLogic<'a> {
         }
         Ok(())
     }
+
+    /// Verify an ED25519 signature given a message and a public key.
+    /// # Cost
+    ///
+    /// `base + write_register_base + write_register_byte * num_bytes + ed25519_verify_base + ed25519_verify_byte * num_bytes`
+    pub fn ed25519_verify(
+        &mut self,
+        sig_len: u64,
+        sig_ptr: u64,
+        msg_len: u64,
+        msg_ptr: u64,
+        pub_key_len: u64,
+        pub_key_ptr: u64,
+    ) -> Result<u32> {
+        use ed25519_dalek::{PublicKey, Signature, Verifier};
+
+        self.gas_counter.pay_base(ed25519_verify_base)?;
+        self.gas_counter.pay_per(ed25519_verify_byte, msg_len + sig_len as u64)?;
```

Comment:
The `msg_len + sig_len` addition can overflow, which would then crash the node. When implementing host functions, be careful about potential overflows. 

There's an extra wrinkle here that, when fetching from a register, the `len` is *not* the lenght of the data, but rather a flag telling "this is a register". 

So, you want to use the resulting len, `signature_array.len()` for accounting for gas, and the same goes for `msg`

Reply:
Really good point!